In [1]:
import numpy as np

# Bevezetés

### Mi a neurális háló?

A neurális hálózat csak egy csomó matematikai egyenlet.

Egy neurális hálózatnak akár több millió állítható paramétere van (súlyok és torzítások), ezért **a neurális hálózatok paraméterekkel rendelkező hatalmas függvényekként viselkednek**. Több millió változóval rendelkező függvény, amit egy probléma megoldására használunk.

**Ennyi neuronokhoz kapcsolódó, összekapcsolt rétegekként elrendezett változóval elképzelhető, hogy léteznek olyan kombinációi e változók értékeinek, amelyek a kívánt kimeneteket eredményezik. *A paraméterek (súly és torzítás) értékek kombinációjának megtalálása jelenti a kihívást*** (ez a betanítás).

A neurális hálózatok végső célja, hogy súlyaikat és torzításaikat (a paramétereket) úgy állítsák be, hogy egy még nem látott bemeneti példára alkalmazva a kívánt kimenetet produkálják. 

Amikor felügyelt gépi tanulási algoritmusokat képezünk, az algoritmusnak bemeneti példákat és a hozzájuk tartozó kívánt kimeneteket mutatunk. 

Az elképzelés az, hogy egy neurális hálózatot sok adatpéldán betanítunk a súlyok és torzítások lassú beállításával, hogy idővel megtaláljuk azt a paramétermilliók kombinációit, ami már elég pontos eredményt ad.

A neurális hálózatok betanításához kiszámítjuk, hogy mennyire téved a modell a hiba (veszteség vagy loss) kiszámítására szolgáló algoritmusok segítségével, és megpróbáljuk frissíteni paramétereiket, hogy sok iteráció során a hálózat fokozatosan egyre kevésbé tévedjen. 


### Bemeneti és kimenetei réteg

A **bemeneti réteg** képviseli a tényleges bemeneti adatokat, például egy kép pixelértékeit vagy egy hőmérséklet-érzékelő adatait. Az adatok jellemzően előfeldolgozottak. (Általában az adatokat úgy dolgozzuk fel előzetesen, hogy közben megőrizzük a jellemzőiket, és az értékek hasonló, 0 és 1 vagy -1 és 1 közötti tartományokban legyenek. Ennek eléréséhez skálázási és normalizálási függvényeket használunk.)

A **kimeneti réteg** az, amit a neurális hálózat visszaad. Osztályozás esetén, ahol a bemenet osztályának előrejelzése a cél, a kimeneti réteg annyi neuront tartalmaz, ahány osztálya van a képzési adathalmaznak, de bináris (két osztály) osztályozás esetén egyetlen kimeneti neuron is lehet. Például a kimeneti réteg két neuronból fog állni: az egyik neuron a "kutyához", a másik pedig a "macskához" kapcsolódik, de lehetne egyetlen kimeneti neuron is, ami a "kutya" vagy a "nem kutya". A legmagasabb pontszámot kapott kimeneti neuron lesz a bemenetként használt kép osztályjóslata.

### Általánosítás

Az egyik fő probléma a túlillesztés, amikor az algoritmus csak a képzési adatokhoz való illeszkedést tanulja meg, ekkor a hálózat lényegében csak "megjegyzi" a képzési adatokat. Ezért "mintán belüli" adatokat használunk a modell képzéséhez, majd "mintán kívüli" adatokat használunk a neurális hálózati modell validálásához. 

**Ezt hívják általánosításnak: az adatok memorizálása helyett az adatokhoz való illeszkedést tanulja meg.** Minden neurális hálózat fő célja az általánosítás, vagyis, hogy korábban soha nem látott adatpéldán pontosan kiadja a kívánt értékeket. 


### Milyen adatokon?

A lineáris adatok nagyon könnyen közelíthetők a neurális hálózatoknál sokkal egyszerűbb gépi tanulási modellekkel. Amit más gépi tanulási algoritmusok nem tudnak közelíteni, azok a nem lineáris adathalmazok. A neurális hálózatok olyan függvényeket készítenek ami matematikailag leírhatatlan sok paraméterrel (görbülettel) rendelkező függvények.

# Neuron felépítése

Minden neuronnak van: 

- bemenet
- minden bemenethez van egy súly
- bias (torzítás)

A bemenetek azok az adatok, amiket átadunk a neuronnak, a súlyok pedig azok az adott bemenethez tartozó paraméterek, amiket később a kívánt eredmények elérése érdekében hangolunk. A bias egy további hangolható érték, de a súlyokkal ellentétben nem kapcsolódik semmilyen bemenethez (maga a neuron tulajdonsága).

A súlyok és a torzítások értékei azok, amiket "betanítunk" (beállítunk a tanulás alatt).

In [2]:
inputs = [1.4, 4.2, 3.8]
weights = [0.2, 0.8, -0.5]
bias = 3

Minden bemenetet megszorzunk a súlyával, összeadjuk az így kapott bemenetek szorzatát majd hozzáadjuk a biast: **így kapjuk meg a neuron kimenetét.**

A kimenet hozzákapcsolódik minden következő réteghez, így **egy pont kimenete minden következő réteg egyik bemenete.**

<img src="kepek/neuron.jpeg" width="400"/>
<img src="kepek/osszeg.jpeg" width="400"/>

In [3]:
output = (inputs[0] * weights[0]) + (inputs[1] * weights[1]) + (inputs[2] * weights[2]) + bias

In [4]:
output

4.74

Numpy megvalósítás:

In [5]:
inputs = np.array([1.4, 4.2, 3.8])
weights = np.array([0.2, 0.8, -0.5])
bias = 3

In [6]:
output = (inputs * weights).sum() + bias

In [7]:
output

4.74

A NumPy `dot` függvény direkt erre van: össze szorozza az elemeket majd összeadja őket

<img src="kepek/dot.jpeg" width="400"/>

In [8]:
np.dot(inputs, weights) + bias

4.74

<br>

# Rétegek

A rétegek neuronok csoportjai. Egy réteg minden neuronja pontosan ugyanazt a bemenetet - a rétegnek adott bemenetet (amely lehet akár a képzési adat, akár az előző réteg kimenete) - veszi fel, de saját súlykészletet és saját biast tartalmaz, és így saját egyedi kimenetet állít elő. 

Teljesen összekapcsolt neurális hálózat: az aktuális réteg minden neuronja kapcsolatban áll az előző réteg minden neuronjával. (Ez a leggyakoribb neurális hálózattípus.)


Három neuron négy inputtal: két layerből álló neurális háló.

(Az output layerek a valóságban mások, mert van egy saját függvényük, de most e nélkül képzeljük el hogy ez a négy az output.)

In [9]:
# A négy input pont bemeneti értékei (a bemeneti érték egy adatsor elemei, nem a hálózat hanem az adat tudlajonsága)
inputs = np.array([1, 2, 3, 2.5])

# Mind a három neuronnak mind a négy bemenetének saját súlya van (ez már a hálózat, azon belül az adott neuron tulajdonsága):
weights1 = np.array([0.2, 0.8, -0.5, 1.0])
weights2 = np.array([0.5, -0.91, 0.26, -0.5])
weights3 = np.array([-0.26, -0.27, 0.17, 0.87])

# mind a négy pontnak saját biasa van (ez is a hálózat, vagyis a neuron tulajdonsága):
b1, b2, b3 = 2, 3, 0.5

<img src="kepek/layers.jpeg" width="400"/>

In [10]:
# kiszámolom mind a három output értékét:
outputs = [np.dot(inputs, weights1) + b1,
           np.dot(inputs, weights2) + b2,
           np.dot(inputs, weights3) + b3
          ]

In [11]:
outputs

[4.8, 1.21, 2.385]

Hogyan tanul ez a rendszer?

A minden réteg bemenete adott, a súlyok és bias változgatásával tanítjuk be a rendszert.

---

### NumPy optimalizálás

1. A súlyokat és a biasokat nem külön változókban tároljuk hanem egy többdimenzós listában:

In [12]:
inputs = np.array([1, 2, 3, 2.5])

weights = np.array([[0.2, 0.8, -0.5, 1.0], 
                    [0.5, -0.91, 0.26, -0.5], 
                    [-0.26, -0.27, 0.17, 0.87]])

biases = np.array([2, 3, 0.5])

2. for loop-al számoljuk ki a dolgokat:

In [13]:
# Python megoldás
outputs = []
for node_weights, node_bias in zip(weights, biases):
    # az inputs konstanas: minden pontnak ugyan az
    # összeszorzom a pontot és a súlyát
    input_x_suly = [weight*input for weight, input in zip(node_weights, inputs)]
    # összeadom az összegeket és hozzáadom a biast
    output = sum(input_x_suly) + node_bias
    outputs.append(output)

In [14]:
outputs

[4.8, 1.21, 2.385]

3. For loop helyett vektor műveletek

- 1D array = Vector - shape(4, )
- 2D array = Matrix (list of vectors) - shape(2, 4) két sor, négy oszlop
- 3D array (list of list of vectors) - shape(3, 2, 4) két dimenzió, minden dimenzióban két sor és négy oszlop

Az `np.dot()` metódus a mátrixot vektorok listájaként kezeli, és mindegyik vektornak a másik vektorral való ponttételét végzi el.

A súlyok minden sorát meg akarom szorozni a bemenet listával:

In [15]:
inputs = np.array([1, 2, 3, 2.5])

weights = np.array([[0.2, 0.8, -0.5, 1.0], 
                    [0.5, -0.91, 0.26, -0.5], 
                    [-0.26, -0.27, 0.17, 0.87]])

biases = np.array([2, 3, 0.5])

A `dot` műveletet elvégezhetjük egy mátrixon: egy mátrix minden során (minden vektorán) elvégez a `dot` műveletet a másik vektorral.

<img src="kepek/dotmatrix.jpeg" width="400"/>

Bemenetnek mindig  mátrixot kell elsőnek és a vektrt másodiknak megadni. (A NumPy-ban mindig az első bemenet haározza meg a műveleti sorrendet: annyi elemet kapun vissza ahány soros a (bemenet) mátrix.)

In [16]:
np.dot(weights, inputs) + biases

array([4.8  , 1.21 , 2.385])

<br>

## Batch of Data

A neurális hálózatok a tanításhoz kötegekben (csoportokban) kapják az adatokat. Ezek az értékek (vektorok) mindegyike egy-egy (megfigyelési) adat, és együttesen egy jellemzőkészlet-példányt alkotnak, amit mintának nevezünk. **Vagyis a batch egy mátrix, bemeneti adatsorok (vektorok) kötege.**

A batch size általában 32 vagy 64. Ennyi mintát számol ki egyszerre és az **ezekre kapott eredménnyel módosítja a függvényt**.

A lényeg az, hogy egyszerre nem egy mintát számolunk ki hanem több mintát. (És majd a tanuló algoritmusba ezzel a négy mintára kiszámolt eredménnyel módosítjuk a függvényt (a neuron éleinek súlyát és biasát)).

Két dolog miatt használjuk ezeket: 
1. **A párhuzamos feldolgozás során gyorsabb a kötegekben történő képzés.**
   
A gépitanulás rengeteg apró számolásból áll: akkor hatékony ha pararell tud futni. GPU több ezer csak azonos utasítást elvégezni képes magjai tökéletesek batch-ek számítására, ezzel a tanítási idő drámaian lecsökken.


3. **A kötegek segítenek az általánosításban a képzés során.**

Ha egyszerre csak egy mintára illeszt (a képzési folyamat egy lépését hajtja végre), akkor nagy valószínűséggel továbbra is az adott egyedi mintára fog illeszkedni, ahelyett, hogy lassan általános, az egész adathalmazra illeszkedő súlyok és torzítások általános módosításait végezné el. A kötegekben történő illesztés nagyobb esélyt ad arra, hogy a súlyok és torzítások értelmesebb módosításait végezze el.

### Inputok

Az inputok összetartozó elemek, például:
- Több szenzor egy időpillanatban rögzített adatai
- Egy kép képkockáinak az értékei
- Egy ember több tulajdonsága

In [17]:
ember1 = [10, 2, 0.87] # pl.: pontszám 8, kedvenc tantárgya: 2, intelligencia (normalizálva): 0.87
ember2 = [8, 3, -0.58]
ember3 = [9, 1, 0.23]

In [18]:
batch = [[10, 2, 0.87], 
         [8, 3, -0.58], 
         [9, 1, 0.23]]

Ebben a mátrixban minden egyes vektor egy mintát jelent, ami egy jellemzőkészletet képvisel (ezeket a listákat jellemzőhalmaz-példányoknak vagy megfigyeléseknek is nevezzük).

Jelen példában három input sorunk van: batch = 3

---

<img src="kepek/layers.jpeg" width="400"/>

A négy bemenet és a három neuronos hálózat megmarad, de most több inputunk van -> **minden inputtal ki kell számolnunk a kimenetet**.

In [19]:
# batch of inputs:
inputs = np.array([[1, 2, 3, 2.5],           # input data 1 (ez látható az ábrán)
                   [2.0, 5.0, -1.0, 2],      # input data 2
                   [-1.5, 2.7, 3.3, -0.8]])  # input data 3

weights = np.array([[0.2, 0.8, -0.5, 1.0],       # 1. neruon súlyai
                    [0.5, -0.91, 0.26, -0.5],    # 2. neuron súlyai 
                    [-0.26, -0.27, 0.17, 0.87]]) # 3. neuron súlyai

biases = np.array([2, 3, 0.5])   # a réteg minden neuronjának torzításai egy listában

In [20]:
for input_sor in inputs:
    outputs = np.dot(weights, input_sor) + biases
    print(outputs)

[4.8   1.21  2.385]
[ 8.9  -1.81  0.2 ]
[1.41  1.051 0.026]


Ezt megoldhatujuk for loop-al is, de az nem hatékony.

### Matrix Product

Nem azt csináljuk, hogy minden input sorral egyesével számoljuk ki az eredményeket hanem az inputokat együtt külldjük át a hálón. **Egy számolással több outputot tudunk kiszámolni a vektorműveletek segítségével.**

Így már két mátrixunk van: súlyok, inputok

A mátrixszorzat egy olyan művelet, amiben van 2 mátrixunk, és **az első mátrix sorainak és a második mátrix oszlopainak minden kombinációjából dot függvényt végzünk**, és egy új, harmadik mátrixot kapunk.

A dot függvény, ha két mátrixon futattjuk:

<img src="kepek/dotmatrixok.jpeg" width="500"/>

(Tehát az első mátrix sorának annyi eleműnek kell lennie mint a második mátrix oszlopának. Tehát csak arra kell oda figyelnünk, hogy annyi súlya legyen egy neuronnak ahány elemből áll egy input, de ez logikailag nem is lehet máshogy.)


Nekünk pont erre van szükségünk: **minden inputtal ki akarjuk számolni az összes neuron kimenetét.**

<img src="kepek/inputsneurons.jpeg" width="700"/>

Hogy ezt a dot függvényel meg tudjuk valósítani a neuronokat (a neuronok súlyait) amik sorokban vannak oszlopokká kell alakítani. Erre való a transzponálás:

<img src="kepek/transpose.jpeg" width="500"/>

(Egy sor vektor transzponálással oszlop vektorrá alakul, és fordítva. **A transzpozíció úgy módosítja a mátrixot, hogy a sorai oszlopokká, az oszlopok pedig sorokká válnak.**)

Összes input kiszámolása az összes neuronon:

In [21]:
np.dot(inputs, weights.T) + biases

array([[ 4.8  ,  1.21 ,  2.385],
       [ 8.9  , -1.81 ,  0.2  ],
       [ 1.41 ,  1.051,  0.026]])

Mindegyik input vektort megszorozzuk minden neuron összes súlyával: az így kapott mátrix az összes neuron kimeneteit tartalmazza. Már csak a bias kell hozzáadni (minden sorhoz ugyan azt a biases vektort, mivel a sorok a három neuron kimenete)

$\left[\begin{array}{ccc}2.8 & -1.79 & 1.885 \\ 6.9 & -4.81 & -0.3 \\ -0.39 & -1.949 & -0.474\end{array}\right]+\left[\begin{array}{lll}2.0 & 3.0 & 0.5\end{array}\right]=\left[\begin{array}{ccc}4.8 & 1.21 & 2.385 \\ 8.9 & -1.51 & 0.2 \\ 1.41 & 1.061 & 0.006\end{array}\right]$


**Három neuronom van (vagyis három oszlopos a kimenet) és a batch is három (vagyis három soros a kimenet).**


<br>

# Még egy layer hozzáadása

A rejtett réteg nem bemeneti vagy kimeneti réteg. A neurális hálózatok akkor lesznek "mélyek", ha 2 vagy több rejtett réteggel rendelkeznek.

Mi azokat az értékeket látjuk amiket a bemeneti rétegnek átadunk és az eredményül kapott adatokat a kimeneti rétegből. Az e végpontok közötti rétegek olyan értékekkel rendelkeznek, amikkel mint felhasználó nem foglalkozunk. De ezek az értékek nagyon fontosak a hálózat tanításához, ugyan is ezek módosításával változik meg a kimeneti réteg eredménye.

Az előző réteg kimenetei mindig a következő réteg bemenetei lesznek. 

A következő rétegnél annyi súlykészletünk lehet, ahányat akarunk (mert bárhány neuronja lehet az új rétegnek), de mindegyik súlykészletnek annyi különálló súlynak kell lennie ahány elemű az előző réteg.

In [22]:
inputs = np.array([[1, 2, 3, 2.5],
                   [2.0, 5.0, -1.0, 2], 
                   [-1.5, 2.7, 3.3, -0.8]])

# Layer 1
weights1 = np.array([[0.2, 0.8, -0.5, 1.0], 
                    [0.5, -0.91, 0.26, -0.5], 
                    [-0.26, -0.27, 0.17, 0.87]])

biases1 = np.array([2, 3, 0.5])


# Layer 2
weights2 = np.array([[0.1, -0.14, 0.5], 
                    [-0.5, 0.12, -0.33], 
                    [-0.44, 0.73, -0.13]])

biases2 = np.array([-1, 2, -0.5])

In [23]:
# az első réteg neuronjaink kimenetei
layer_1_outputs = np.dot(inputs, weights1.T) + biases1

In [24]:
# a második réteg neuronjaink kimenetei
layer_2_outputs = np.dot(layer_1_outputs, weights2.T) + biases2 # a bemenet az előző réteg kimenetet

In [25]:
layer_2_outputs

array([[ 0.5031 , -1.04185, -2.03875],
       [ 0.2434 , -2.7332 , -5.7633 ],
       [-0.99314,  1.41254, -0.35655]])

---

## Dense Layer Class

Az előzőkben úgynevezett sűrű vagy teljesen összekapcsolt réteget használtuk. Az ilyen réteget dense vagy fully-connected layer-nek nevezik.

Amikor egy modellen az elejétől a végéig átvezetjük az adatokat, azt forward pass-nak nevezzük.


### Inicializálási értékek

**A torzítást nullának szokás inicializálni.**

De néha szükséges lehet a torzításokat nem nullával inicializálni. Például a halott neuronok esetében. Lehetséges, hogy a `súlyok*bemenetek+bias` nem érik el a lépésfüggvény küszöbértékét, ami azt jelenti, hogy a neuron 0-t fog kiadni. Ez nem nagy probléma, de problémává válik, ha ez minden bemeneti mintánál megtörténik ezzel a neuronnal. Ekkor tehát ennek a neuronnak a 0 kimenete egy másik neuron bemenete. Minden nullával szorzott súly nulla lesz. A 0-t kibocsátó neuronok növekvő számával a következő neuronok több bemenete fogja megkapni ezeket a 0-kat, így a hálózat lényegében "halott".

**A súlyokat véletlenszerűen inicializáljuk egy modellhez.** (De ha egy előre betanított modellt szeretnénk betölteni, akkor a paramétereket azzal inicializáljuk, amivel az előre betanított modell befejezte. Ez azt is jelenti, hogy amikor elmentünk egy beatított modellt csak a súly és a bias értékeket kell elmenteni.)

Az `np.random.randn` egy Gauss-eloszlást állít elő 0 átlaggal és 1 varianciával: pozitív és negatív véletlen számokat fog generálni, amiknek középpontja 0, és az átlagértékük is közel 0. Ezt a Gauss-eloszlást a súlyokhoz 001-gyel fogjuk megszorozni, hogy néhány nagyságrenddel kisebb számokat generáljunk. (E nélkül a modellnek több időbe fog telni az adatokhoz való illesztés a képzési folyamat során mivel a kiindulási értékek aránytalanul nagyok lesznek a frissítésekhez képest.)

**Az ötlet az, hogy a modellt olyan nem nulla értékekkel indítsuk, amik elég kicsik ahhoz, hogy ne befolyásolják a képzést.** Így egy csomó értékkel kezdhetünk dolgozni, de remélhetőleg egyik sem túl nagy vagy 0. (Elég kicsi, hogy ne befolyásolja az eredményt, de mégis működik mert nem nulla.)

Amikor elkezdjük betanítani a modellt, akkor azt várjuk, hogy a random számok elkezdenek egyre kevésbé randomak lenni. És ahogy elkezdenek változni a számok akkor a tartomány kitágul (mint mikor a jel kimagaslik a zajból). Tehát nem akarjuk hogy a random számok (zaj) értéke is nagy legyen.

---

Az input layert az adat szabja meg, az egyes adatpontoknak hány tulajdonsága van. Az output layert a feladat szabja meg: hány lehetőség közül lehet választani (kutya vagy macska). Amit mi megadhatunk azok a köztes layerek amit Hidden Layers-nek neveznek.

Amikor létrehozok egy neurális hálózatot **létre kell hoznom a súlyokat és bias-okat. Matematikailag csak ebből áll a hálózat.**

A súlyok értéke -1 és +1 között tud lenni. A hálózatok használatának 0. lépése, hogy normalizálni kell a bemenetet (az arány megmarad, de a tartomány -1 és +1).

**A neuronoknak van egy tüzelési küszöbértékük, a kimenetük nulla ha nem érik el ezt a megadott határt.**

A -1 +1 tartományban a kimenet bias nélkül annyira kicsi hogy nem éri a tüzelési küszöböt. A bias-al szabályozzuk, hogy tüzeljen e és honnan. Értéke 0 és 10 között mozog általában. Ha nulla akkor biztos hogy nem fog tüzelni a neuron bármennyire nagy is legyen a súly és a bemenet.

*[Ez egy önellentmondás a fentiekhez képest. De akkor melyik a helyes??]*

---

Egy neurális hálózat létrehozásakor két dolgot kell tudnunk:
1. Input layer nagysága (hány elemből áll egy bemenet) (az adatokból tudjuk)
2. Hidden Layerek száma

In [26]:
np.random.seed(0)

In [27]:
class Layer_Dense:
    def __init__(self, inputs_db, neurons_db):
        # a súlyoknak fordított a shape-el hozzuk létre, mert így nem kell majd transpozeolni
        self.weight = np.random.randn(inputs_db, neurons_db) * 0.01 
        self.biases = np.zeros(neurons_db)

    def forward(self, inputs):
        self.output = np.dot(inputs, self.weight) + self.biases # nem kell transpose

In [28]:
layer1 = Layer_Dense(4, 3)
layer2 = Layer_Dense(3, 3)
layer3 = Layer_Dense(3, 2)

Arra az egyre kell nagyon figyelni, hogy a következő layernek annyi bemenetének kell lennie, ahány kimenete van az előző layernek. Az output db-nak (3) meg kell eggyeznie a következő layer input db-janak (3)

In [29]:
# ez már az oszlop vektorokat tartalmazza: 3 neuron: három oszlop, 4 input: négy eleműek az oszlopok -> minden inputhoz egy sor 
layer1.weight 

array([[ 0.01764052,  0.00400157,  0.00978738],
       [ 0.02240893,  0.01867558, -0.00977278],
       [ 0.00950088, -0.00151357, -0.00103219],
       [ 0.00410599,  0.00144044,  0.01454274]])

Most, hogy megvannak a layereink csak végig kell nyomni rajta az adatot.

In [30]:
# X = data inputs
X = np.array([[1, 2, 3, 2.5],
              [2.0, 5.0, -1.0, 2], 
              [-1.5, 2.7, 3.3, -0.8]])

In [31]:
layer1.forward(X) # ezzel létrejönek az objektumon belül az output értékek

In [32]:
layer1.output

array([[ 0.101226  ,  0.0404131 ,  0.02350209],
       [ 0.14603679,  0.10577549,  0.00082852],
       [ 0.06211146,  0.03827457, -0.05610798]])

Most ezeket az adatokat kell továbbadni a második layer inputjának:

In [33]:
layer2.forward(layer1.output)

In [34]:
layer2.output

array([[ 0.00097879,  0.00052624, -0.00023361],
       [ 0.00146693,  0.00175098,  0.00041004],
       [ 0.00042475,  0.00112664,  0.0016296 ]])

In [35]:
layer3.forward(layer2.output)

In [36]:
layer3.output

array([[ 5.88957871e-06,  2.02985089e-05],
       [-9.37057260e-06,  5.26113685e-05],
       [-2.92856209e-05,  2.99893735e-05]])